cmd/soroban-rpc/internal/events/events.go

```diff
@@ -0,0 +1,219 @@
+// -1 is returned if c is less than other.
+func (c Cursor) Cmp(other Cursor) int {
+	if c.Ledger == other.Ledger {
+		if c.Tx == other.Tx {
+			return cmp(c.Event, other.Event)
+		}
+		return cmp(c.Tx, other.Tx)
+	}
+	return cmp(c.Ledger, other.Ledger)
+}
+
+// MemoryStore is an in-memory store of soroban events.
+type MemoryStore struct {
+	lock            sync.RWMutex
+	events          []event
+	length          int
+	start           int
+	retentionWindow uint32
+}
+
+// NewMemoryStore creates a new MemoryStore populated by the given ledgers.
+func NewMemoryStore(passPhrase string, ledgers []xdr.LedgerCloseMeta, retentionWindow uint32) (*MemoryStore, error) {
+	var events []event
+	for _, ledger := range ledgers {
+		reader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(passPhrase, ledger)
+		if err != nil {
+			return nil, err
+		}
+		ledgerEvents, err := readEvents(reader)
+		if err != nil {
+			return nil, err
+		}
+		events = append(events, ledgerEvents...)
+	}
+
+	return &MemoryStore{
+		events:          events,
+		retentionWindow: retentionWindow,
+		length:          len(events),
+	}, nil
+}
+
+// Scan applies f on all the events occurring in the range of [start, end).
+// The events are processed in sorted ascending Cursor order.
+// If f returns false, the scan terminates early (f will not be applied on
+// remaining events in the range).
+func (m *MemoryStore) Scan(start, end Cursor, f func(Cursor, xdr.ContractEvent) bool) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	i := sort.Search(m.length, func(i int) bool {
+		index := (m.start + i) % m.length
+		entry := m.events[index]
+		return entry.cursor.Cmp(start) < 0
+	}) + 1
+
+	for ; i < m.length; i++ {
+		index := (m.start + i) % m.length
+		entry := m.events[index]
+		if entry.cursor.Cmp(end) >= 0 {
+			break
+		}
+		if !f(entry.cursor, entry.contents) {
+			break
+		}
+	}
+}
+
+// IngestEvents adds new events from the given ledger into the store.
+// As a side effect, events which fall outside the retention window are
+// removed from the store.
+func (m *MemoryStore) IngestEvents(txReader *ingest.LedgerTransactionReader) error {
+	events, err := readEvents(txReader)
+	if err != nil {
+		return err
+	}
+	if len(events) == 0 {
+		return nil
+	}
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	m.trim(txReader.GetSequence())
+	m.maybeResize(len(events))
+	m.append(events)
+	return nil
+}
+
+func readEvents(txReader *ingest.LedgerTransactionReader) ([]event, error) {
+	var events []event
+	sequence := txReader.GetSequence()
+	for {
+		tx, err := txReader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		// TODO : add function in ingest.LedgerTransaction to obtain operation events
+		txMeta, ok := tx.UnsafeMeta.GetV3()
+		if !ok {
+			continue
+		}
+		for opIndex, op := range txMeta.Events {
+			if len(op.Events) == 0 {
+				continue
+			}
+			for eventIndex, opEvent := range op.Events {
+				events = append(events, event{
```

Comment:
is it worthwhile to offer an external config to be used here as event topic filters? as this is crunching events for the whole network, maybe rpc hosting use cases will only be interested in events for their contract or known topics, etc?